codec-dns/src/main/java/io/netty/handler/codec/dns/DnsMessage.java

```diff
@@ -15,225 +15,138 @@
  */
 package io.netty.handler.codec.dns;
 
-import io.netty.util.AbstractReferenceCounted;
-import io.netty.util.ReferenceCountUtil;
+import io.netty.util.ReferenceCounted;
 
-import java.util.Collections;
-import java.util.LinkedList;
 import java.util.List;
 
 /**
  * The message super-class which contains core information concerning DNS
  * packets, both outgoing and incoming.
  */
-public abstract class DnsMessage extends AbstractReferenceCounted {
-
-    private List<DnsQuestion> questions;
-    private List<DnsResource> answers;
-    private List<DnsResource> authority;
-    private List<DnsResource> additional;
-
-    private final DnsHeader header;
-
-    // Only allow to extend from same package
-    DnsMessage(int id) {
-        header = newHeader(id);
-    }
+public interface DnsMessage extends ReferenceCounted {
```

Comment:
Overall comment after looking at the resulting interface:  Setters/mutators do not belong in API - some [supporting documentation here](https://books.google.com/books?id=VXKBgiD57lwC&pg=PA74&lpg=PA74&dq=setters+do+not+belong+in+api&source=bl&ots=iKQyoCa2wv&sig=oQpuEFC8x40rITm_IldJVUzLHsg&hl=en&sa=X&ei=TR8HVfmUFMKrgwSZnICQDQ&ved=0CB4Q6AEwAA#v=onepage&q=setters%20do%20not%20belong%20in%20api&f=false).

It is perfectly reasonable for an implementation to simply passing immutable DNS objects to a client, where the return value from one of the related getters is computed on the fly, and having a setter makes no sense.  Setters are fine as protected methods in base classes that make it easier to implement things, but if they are built into the API, they

 * Increase the surface area of the API, meaning
    * More conceptual overhead
    * Higher chance of breaking changes
 * Make immutable implementations awkward or impossible
 * Psychologically, if they are there, they are an invitation for clients to call them, whether they should or not